runtime-modules/bounty/src/lib.rs

```diff
@@ -1114,93 +1285,91 @@ decl_module! {
+            Self::deposit_event(RawEvent::WorkSubmitted(bounty_id, entry_id, member_id, work_data));
         }
 
-        /// Submit work for a bounty.
+        /// end bounty working period.
         /// # <weight>
         ///
         /// ## weight
-        /// `O (N)`
-        /// - `N` is the work_data length,
+        /// `O (1)`
         /// - db:
         ///    - `O(1)` doesn't depend on the state or parameters
         /// # </weight>
-        #[weight =  WeightInfoBounty::<T>::submit_work(work_data.len().saturated_into())]
-        pub fn submit_work(
+        #[weight = WeightInfoBounty::<T>::end_working_period()]
+        pub fn end_working_period(
             origin,
-            member_id: MemberId<T>,
             bounty_id: T::BountyId,
-            entry_id: T::EntryId,
-            work_data: Vec<u8>
         ) {
-            T::Membership::ensure_member_controller_account_origin(origin, member_id)?;
-
             let bounty = Self::ensure_bounty_exists(&bounty_id)?;
+            let current_oracle = bounty.creation_params.oracle.clone();
+
+            //Checks if the function caller (origin) is current oracle
+            BountyActorManager::<T>::ensure_bounty_actor_manager(
+                origin,
+                current_oracle.clone(),
+            )?;
 
             let current_bounty_stage = Self::get_bounty_stage(&bounty);
 
             Self::ensure_bounty_stage(current_bounty_stage, BountyStage::WorkSubmission)?;
 
-            let entry = Self::ensure_work_entry_exists(&bounty_id, &entry_id)?;
-
-            Self::ensure_work_entry_ownership(&entry, &member_id)?;
-
             //
             // == MUTATION SAFE ==
             //
 
-            // Update entry
-            <Entries<T>>::mutate(bounty_id, entry_id, |entry| {
-                entry.work_submitted = true;
-            });
-
-            let new_milestone = Self::get_bounty_milestone_on_work_submitting(&bounty);
-
-            // Update bounty record.
             <Bounties<T>>::mutate(bounty_id, |bounty| {
-                bounty.milestone = new_milestone;
+                bounty.milestone = BountyMilestone::WorkSubmitted;
             });
-
-            Self::deposit_event(RawEvent::WorkSubmitted(bounty_id, entry_id, member_id, work_data));
+            Self::deposit_event(RawEvent::WorkSubmissionPeriodEnded(bounty_id, current_oracle));
         }
 
-        /// Submits an oracle judgment for a bounty.
+        /// Submits an oracle judgment for a bounty, slashing the entries rejected
+        /// by an arbitrary percentage and rewarding the winners by an arbitrary amount
+        /// (not surpassing the total fund amount)
         /// # <weight>
         ///
         /// ## weight
-        /// `O (N)`
-        /// - `N` is the work_data length,
+        /// `O (N x M + Z)`
+        /// - `N` is judgment,
+        /// - `M` is action_justification (inside OracleJudgment),
+        /// - `Z` is rationale
```

Comment:
```suggestion
        /// `O (N + M + Z)`
        /// - `N` is number of judgment entries,
        /// - `M` is the sum of all action_justification lengths (inside OracleJudgment),
        /// - `Z` is rationale length
```

I think you should also adjust the benchmarks to reflect this, because currently you're passing the sum of the length of all `action_justification`s as a parameter to the weight function, but the benchmarks effectively multiply the impact of this parameter by `MAX_WORK_ENTRIES` (because there you create one `action_justification` of length `k` per each entry), which means the user will be overpaying a lot.

Reply:
Could you be more clear please, about what you want me to do in benchmarks?

The weight should be the N x M + Z in case of all rejected and N + Z in case of all successful

Reply:
So in `submit_oracle_judgment_by_council_all_rejected` and `submit_oracle_judgment_by_member_all_rejected` you have a parameter `k` and in `submit_oracle_judgment_weight` you have `justification_length` which represents the same parameter.

In `submit_oracle_judgment_weight`, the `justification_length` is **the sum of all `action_justification.len()` in the `judgment_map`**

In `submit_oracle_judgment_by_council_all_rejected` and `submit_oracle_judgment_by_member_all_rejected` benchmarks, `k` is the length of **each `action_justification` in the `judgement` map (so the sum of all `action_justification.len()`s is `k * i`)**

So `k` and `justification_length` mean 2 different things, while they should mean the same one.
One way to solve this is to make `k` in `submit_oracle_judgment_by_council_all_rejected` and `submit_oracle_judgment_by_member_all_rejected` actually represent the sum of all `action_justification.len()`s, for example:

```diff
        let judgment = (0..i).into_iter().map(|it|{
            let entry_id = announce_entry_and_submit_work::<T>(&bounty_id, it);
            let judgment = OracleWorkEntryJudgment::Rejected {
                slashing_share: Perbill::from_percent(50),
-               action_justification: action_justification.clone(),
+               action_justification: match it == 0 {
+                   true => vec![0u8].repeat((k / i + k % i) as usize),
+                   false => vec![0u8].repeat((k / i) as usize)
+               },
            };
            (entry_id, judgment)
        }).collect::<BTreeMap<_, _>>();
```

Another issue here is I don't think `all_rejected` and `all_winners` actually represent good cases for benchmarking, because:
- `all_winners` doesn't take into account `justification_length`
- `all_rejected` doesn't take into account the actual number of winners

So imagine a scenario where there is 1 `Rejected` entry with a huge `justification_length` (for example, 3 megabytes) and 100 `Winner` entires in the `judgment` (so total of 101 entries). In that case:
- `all_winners` benchmark may give an underestimated value, because it won't take into account the huge impact that `justification_length` of the single `Rejected` entry would have on the weight
- `all_rejected` benchmark may also give an underestimated value, because it won't take into account the huge impact that processing 100 `Winner`s would have (I'm assuming processing a `Winner` is of higher complexity, because it involves a transaction, while processing rejected entry involves only slashing)

Therefore **I think that the number of `Winner`/`Rejected` entries should be the weight function parameters.**